File: three_sum.sol.py
from typing import List
from collections import defaultdict 


class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        count_dict = defaultdict(int)
        answer = []
        for i, v in enumerate(nums):
            count_dict[v] += 1
        num = list(sorted(count_dict.keys()))
        for ind_i, i in enumerate(num):
            if i > 0:
                break
            for j in num[ind_i:]:
                target = 0 - i - j
                if target < j:
                    break
                if i == j and count_dict[i] <= 1:
                    continue
                if j == target and count_dict[j] <= 1:
                    continue
                if i == j == target and count_dict[i] <= 2:
                    continue
                if not count_dict.get(target):
                    continue
                answer.append([i, j, target])
        return answer

File: three_sum.sol.java
import java.util.List;
import java.util.Arrays;
import java.util.LinkedList;


class Solution {
    public List<int[]> threeSum(int[] nums) {
        Arrays.sort(nums);
        List<int[]> ans = new LinkedList<>(); 

        for (int i = 0; i < nums.length-2; i++) {
            int left = i + 1;
            int right = nums.length - 1;
            int target = 0 - nums[i];

            while (left < right) {
                if (nums[left] + nums[right] == target) {
                    ans.add(new int[]{nums[i], nums[left], nums[right]});
                    
                    left++; 
                    right--;
                } 
                else if (nums[left] + nums[right] < target) ++left;
                else --right;
           }
            
        }
        return ans;
    }
}
